File: python/data_structures/queue.py
"""
Queue Implementation
Time Complexity: O(1) for all operations
Space Complexity: O(n)
"""

from collections import deque

class Queue:
    """Queue data structure implementation using deque"""
    
    def __init__(self):
        """Initialize empty queue"""
        self.data = deque()
    
    def enqueue(self, value):
        """
        Enqueue element to queue
        
        Args:
            value: Element to enqueue
        """
        self.data.append(value)
    
    def dequeue(self):
        """
        Dequeue element from queue
        
        Returns:
            Front element, or None if queue is empty
        """
        if self.is_empty():
            print("Queue underflow!")
            return None
        return self.data.popleft()
    
    def front(self):
        """
        Get front element without removing it
        
        Returns:
            Front element, or None if queue is empty
        """
        if self.is_empty():
            print("Queue is empty!")
            return None
        return self.data[0]
    
    def is_empty(self):
        """
        Check if queue is empty
        
        Returns:
            True if queue is empty, False otherwise
        """
        return len(self.data) == 0
    
    def size(self):
        """
        Get queue size
        
        Returns:
            Number of elements in queue
        """
        return len(self.data)
    
    def display(self):
        """Display queue contents"""
        if self.is_empty():
            print("Queue is empty")
            return
        print("Queue:", " ".join(map(str, self.data)))

if __name__ == "__main__":
    q = Queue()
    
    q.enqueue(10)
    q.enqueue(20)
    q.enqueue(30)
    
    print("After enqueuing 10, 20, 30:")
    q.display()
    
    print(f"Front element: {q.front()}")
    print(f"Queue size: {q.size()}")
    
    print(f"Dequeued element: {q.dequeue()}")
    q.display()

File: cpp/data_structures/queue.cpp
#include <iostream>
#include <queue>
using namespace std;

// Queue Implementation using STL
// Time Complexity: O(1) for all operations
// Space Complexity: O(n)

class Queue {
private:
    queue<int> data;
    
public:
    // Enqueue element to queue
    void enqueue(int value) {
        data.push(value);
    }
    
    // Dequeue element from queue
    int dequeue() {
        if (isEmpty()) {
            cout << "Queue underflow!" << endl;
            return -1;
        }
        int value = data.front();
        data.pop();
        return value;
    }
    
    // Get front element
    int front() {
        if (isEmpty()) {
            cout << "Queue is empty!" << endl;
            return -1;
        }
        return data.front();
    }
    
    // Check if queue is empty
    bool isEmpty() {
        return data.empty();
    }
    
    // Get queue size
    int size() {
        return data.size();
    }
    
    // Display queue (note: STL queue doesn't support iteration)
    void display() {
        if (isEmpty()) {
            cout << "Queue is empty" << endl;
            return;
        }
        queue<int> temp = data;
        cout << "Queue: ";
        while (!temp.empty()) {
            cout << temp.front() << " ";
            temp.pop();
        }
        cout << endl;
    }
};

int main() {
    Queue q;
    
    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);
    
    cout << "After enqueuing 10, 20, 30:" << endl;
    q.display();
    
    cout << "Front element: " << q.front() << endl;
    cout << "Queue size: " << q.size() << endl;
    
    cout << "Dequeued element: " << q.dequeue() << endl;
    q.display();
    
    return 0;
}
